(Towards 1338) Implementation of OpenMP taskwait directive

LonelyCat124 opened the pull request:
Waiting on #1358. Will rebase to master once that is merged.

From discussion with Rupert we chose to separate the implementation of the directive and transformation, as the transformation is more complex.

All the taskwait directive does at the moment is create relevant code, and check that it is contained within a Parallel Region. It could be more restrictive (and also require to be inside a Serial region) however I think that is not required by OpenMP standard at all, and potentially a valid usecase as is.

The taskwait transformation (future PR) will handle other checks, and will be applied to one or more parallel loops, and handle computation of dependencies etc. to put the minimum number of taskwait directives into that region of code. It is planned that the transformation will spit out an exception when applied to a set of loops without tasks, though this can safely be ignored.

LonelyCat124 commented:
Ok, this should be good to go now. This is the last PR I need before I attempt to create a recipe for the NemoLite2D using `taskloop` and `taskwait`.

arporter reviewed (changes requested):
Nice self-contained PR.
All tests and examples are OK with compilation. (Although we don't actually support compilation testing with OpenMP enabled now that I come to think about it - see #314.)
I suggest that a new example should be added as part of completing #1338 (i.e. not in this PR).
The new node type inherits its colour, parent-child relationship restrictions and copying behaviour from the parent class (OMPDirective) which I think is OK.
Reference guide builds fine. I don't think any other documentation needs updating at this stage.
Just a few minor things to address.

LonelyCat124 commented on src/psyclone/tests/psyir/nodes/omp_directives_test.py:
```diff
@@ -471,6 +472,88 @@ def test_omp_master_nested_validate_global_constraints(monkeypatch):
+
+    schedule.addchild(taskwait, 1)
+    taskloop.apply(schedule.children[0])
+    master.apply([schedule.children[0], schedule.children[1]])
```
Aha, I htink i had tried this but it wasn't working because i got numbering wrong (I did [0:1] I think).

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -112,6 +112,107 @@ def _get_reductions_list(self, reduction_type):
+        self.validate_global_constraints()
+
+        # Generate the code for this Directive
+        parent.add(DirectiveGen(parent, "omp", "begin", "taskwait",
```
Done

LonelyCat124 commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -112,6 +112,107 @@ def _get_reductions_list(self, reduction_type):
+    :type parent: :py:class:`psyclone.psyir.nodes.Node`
+
+    '''
+    def __init__(self, parent=None):
```
From discussion on teams, removed the constructor, and added a `validate_child` static method to forbid children.

LonelyCat124 replied:
Ok - this seems more complex that just that. Andy recommended a new childless Schedule directive , so I will implement that now in a new PR

LonelyCat124 replied:
I implemented that now in PR after discussion. I'm waiting to see coverage on the new directive code now.

LonelyCat124 commented on src/psyclone/tests/psyir/nodes/omp_directives_test.py:
```diff
@@ -471,6 +472,88 @@ def test_omp_master_nested_validate_global_constraints(monkeypatch):
+    taskloop.apply(schedule.children[0])
+    master.apply([schedule.children[0], schedule.children[1]])
+    parallel.apply(schedule.children[0])
+    goceantrans = GOceanExtractTrans()
```
The test fails without it so I think so.

arporter replied:
Oh yes, now that I take a proper look this test is there because things like ExtractNode still use the old gen_code way of creating code rather than using a PSyIR backend.

LonelyCat124 commented:
Ok, @arporter I think I fixed the issues. The one thing I'm not quite sure I understand from our conversation is the things about ` ACCEnterData`. I think what you meant is that this should be a childless directive as well. Towards that end I implemented an `ACCChildlessDirective` class (which currently has no tests as it has no subclass). I tried changing `ACCEnterDataDirective` to be a child of that class but then lots of tests started failing, so I'm not quite sure if/how to do this.

arporter commented:
Sorry for the confusion @LonelyCat124. I looked last night and was surprised to see that ACCEnterData is not childless - it too has a Schedule (you can see this if you `cd examples/gocean/eg4; make openacc`). I then looked at @sergisiso latest PR (#1282) where he's adding support for `!$acc update {host,device}` which ought to be childless and again, it too has a Schedule as a child. It therefore feels like we need a bit of a design discussion about this but I don't know when Sergi is back from leave (if he is on leave)?

LonelyCat124 commented:
Ok, updated this as discussed. I think everything is sorted. I might have missed things due to the progress and reversion of the childless branch. I did end up having to force push to revert those changes on this branch, so hopefully that didn't break anything.

arporter reviewed (changes requested):
All looking good now.
Tests and examples still all OK.
Understanding the need for the GOExtractTrans made me realise that we're missing a test that the new node works OK with the PSyIR backend - please could you add/modify a test in `tests/psyir/backend/psyir_openmp_test.py` to check this.
Please could you also bring the branch up-to-date with master.
Once that's done then this is ready to go.

LonelyCat124 commented:
Ok, test added and master merged in

arporter commented on src/psyclone/tests/psyir/backend/psyir_openmp_test.py:
```diff
@@ -245,3 +246,39 @@ def test_gocean_omp_do():
+    _, invoke = get_invoke("single_invoke.f90", "gocean1.0",
+                           idx=0, dist_mem=False)
+    omp = OMPTaskloopTrans()
+    _, _ = omp.apply(invoke.schedule[0])
```
I think you also need to manually add a TaskWait node to this schedule?

LonelyCat124 replied:
Yep sorry, brain tired and wrote the test for the wrong Task directive. I fixed this now, the test covers both of them as it was hard for me to work out a standalone test for just TaskWait

arporter replied:
Thanks.

arporter reviewed (changes requested):
Almost - your new test covers OMPTaskLoop but not the new directive added in this PR :-) (Unless I've missed something...)

arporter commented:
Coverage is good.
All tests and examples are OK with compilation.
pycodestyle is happy.
All docs build OK.
Will proceed to merge...

arporter reviewed (approved):
All good now - as commented above.
